File: house_robber/house_robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        cache = {
            len(nums): 0,
            len(nums)+1: 0,
            len(nums)+2: 0
        }
        return self.recursion_helper(
            0,
            nums,
            cache
        )
        
    
    def recursion_helper(self, index: int, nums: List[int], cache: dict[int, int]) -> int:
        if index in cache:
            return cache[index]

        skip = self.recursion_helper(index+1, nums, cache)
        take = nums[index] + self.recursion_helper(index+2, nums, cache)

        cache[index] = max(skip, take)
        return cache[index]
        

File: house_robber/house_robber.go
package houserobber

func rob(nums []int) int {
	cache := map[int]int{
		len(nums):     0,
		len(nums) + 1: 0,
		len(nums) + 2: 0,
	}
	return recursionHelper(
		0,
		nums,
		cache,
	)
}

func recursionHelper(index int, nums []int, cache map[int]int) int {
	if value, existing := cache[index]; existing {
		return value
	}

	skip := recursionHelper(
		index+1,
		nums,
		cache,
	)

	take := nums[index] + recursionHelper(
		index+2,
		nums,
		cache,
	)

	var maxValue int
	if skip > take {
		maxValue = skip
	} else {
		maxValue = take
	}

	cache[index] = maxValue
	return cache[index]
}
